src/python/WMCore/MicroService/Unified/MSMonitor.py

```diff
@@ -17,123 +21,170 @@ class MSMonitor(MSCore):
+        if not campaigns:
+            self.logger.warning("Failed to fetch the campaign configurations")
+        else:
+            # let's make campaign look-up easier and more efficient
+            self.campaigns = {}
+            for camp in campaigns:
+                if 'campaignName' not in camp:
+                    self.logger.warning(
+                            'No campaignName attribute in campaign dict: %s',
+                            json.dumps(camp))
+                    continue
+                self.campaigns[camp['campaignName']] = camp
+            return True
+        return False
+
     def execute(self, reqStatus):
         """
-        Executes the MS monitoring logic
+        Executes the MS monitoring logic, see
+        https://github.com/dmwm/WMCore/wiki/ReqMgr2-MicroService-Monitor
+
         :param reqStatus: request statue to process
         :return:
         """
+        if not self.updateCaches():
+            # then wait until the next cycle
+            msg = "Failed to fetch data from one of the data sources. Retrying again in the next cycle"
+            self.logger.error(msg)
+            return
+
         try:
             # get requests from ReqMgr2 data-service for given statue
             # here with detail=False we get back list of records
             requests = self.reqmgr2.getRequestByStatus([reqStatus], detail=False)
             self.logger.debug('+++ monit found %s requests in %s state',
                               len(requests), reqStatus)
 
-            # FIXME: completion threshold should come either from unified
-            # or campaign configuration, e.g.
-            # thr = self.unifiedConfig().get('completedThreshold', 100)
-            thr = 100
-
-            requestStatuses = {}  # keep track of request statuses
+            # get list of all transfer records
+            transferRecords = [d for d in self.getTransferInfo('ALL_DOCS')]
+            # keep track of completed workflows
+            completedWorkflows = []
+            # keep track of request and their new statuses
+            reqStatus = {}
+            # main logic
             for reqName in requests:
-                req = {'name': reqName, 'reqStatus': reqStatus}
-                self.logger.debug("+++ request %s", req)
-                # obtain status records from couchdb for given request
-                transferRecords = self.getTransferRecords(reqName)
-                if not transferRecords:
+                self.logger.debug("+++ request %s", reqName)
+                # for us workflow name is the same as request name
+                wname = reqName
+                # obtain transfer records for our request
+                transfers = []
+                for rec in transferRecords:
+                    if reqName == rec['workflowName']:
+                        transfers = rec['transfers']
+                        break
+                if not transfers:
```

Comment:
This has no distinction between a transfer document without any transfers FROM the lack of a transfer document for a given workflow. We know every workflow should have an associated transfer document, but in case it doesn't, the very least we need to do is to throw an ERROR in the logs.